sktime/transformations/compose.py

```diff
@@ -0,0 +1,464 @@
+# -*- coding: utf-8 -*-
+"""Meta-transformers for building composite transformers."""
+# copyright: sktime developers, BSD-3-Clause License (see LICENSE file)
+
+from sklearn.base import clone
+
+from sktime.base import _HeterogenousMetaEstimator
+from sktime.transformations.base import BaseTransformer
+
+__author__ = ["fkiraly"]
+__all__ = ["TransformerPipeline"]
+
+
+class TransformerPipeline(BaseTransformer, _HeterogenousMetaEstimator):
+    """Pipeline of transformers compositor.
+
+    The `TransformerPipeline` compositor allows to chain transformers.
+    The pipeline is constructed with a list of sktime transformers,
+        i.e., estimators following the BaseTransformer interface.
+    The list can be unnamed - a simple list of transformers -
+        or string named - a list of pairs of string, estimator.
+
+    For a list of transformers `trafo1`, `trafo2`, ..., `trafoN`,
+        the pipeline behaves as follows:
+    `fit` - changes state by running `trafo1.fit_transform`, `trafo2.fit_transform`, etc
+        sequentially, with `trafo[i]` receiving the output of `trafo[i-1]`
+    `transform` - result is of executing `trafo1.transform`, `trafo2.transform`, etc
+        with `trafo[i].transform` input = output of `trafo[i-1].transform`,
+        and returning the output of `trafoN.transform`
+    `inverse_transform` - result is of executing `trafo[i].inverse_transform`,
+        with `trafo[i].inverse_transform` input = output `trafo[i-1].inverse_transform`,
+        and returning the output of `trafoN.inverse_transform`
+    `update` - changes state by chaining `trafo1.update`, `trafo1.transform`,
+        `trafo2.update`, `trafo2.transform`, ..., `trafoN.update`,
+        where `trafo[i].update` and `trafo[i].transform` receive as input
+            the output of `trafo[i-1].transform`
+
+    `get_params`, `set_params` uses `sklearn` compatible nesting interface
+        if list is unnamed, names are generated as names of classes
+        if names are non-unique, `f"_{str(i)}"` is appended to each name string
+            where `i` is the total count of occurrence of a non-unique string
+            inside the list of names leading up to it (inclusive)
+
+    `TransformerPipeline` can also be created by using the magic multiplication
+        on any transformer, i.e., any estimator inheriting from `BaseTransformer`
+            for instance, `my_trafo1 * my_trafo2 * my_trafo3`
+            will result in the same object as  obtained from the constructor
+            `TransformerPipeline([my_trafo1, my_trafo2, my_trafo3])`
+        magic multiplication can also be used with (str, transformer) pairs,
+            as long as one element in the chain is a transformer
+
+    Parameters
+    ----------
+    transformer : list of sktime transformers, or
+        list of tuples (str, transformer) of sktime transformers
```

Comment:
is `List` also allowed in `sklearn.Pipeline`? I thought its required to be `List[Tuple]`? So might stick to what `sklearn` uses?

Reply:
yes it says `stepslist of tuple`: https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html

Reply:
it extends the interface. Why always the hassle to name your estimators if you can type it more quickly?
Also, the syntax `mytrafo1 * mytrafo2` would not work without.

Reply:
but without giving a name, you cant tune it as you need the name to do `mytrafo__method="trend"`.

Reply:
it auto-generates names which you can look up in `transformers_`.
In the simplest case, the name is just the class name (if they are unique), otherwise `_0`, `_1` are appended etc.

Reply:
(& the names are unique for the entire pipeline - if you add more elements, and accidentally create a name clas, it is auto-reolved)

Reply:
```
    transformers_ : list of tuples (str, transformer) of sktime transformers
        clones of transformers in `transformers` which are fitted in the pipeline
 -->    is always in (str, transformer) format, even if transformers is just a list
        strings not passed in transformers are unique generated strings
        i-th transformer in `transformers_` is clone of i-th in `transformers`
```